File: src/main/java/tools/important/tankslua/gui/ToggleButton.java
package tools.important.tankslua.gui;

import tanks.gui.Button;
import tanks.gui.screen.ScreenOptions;

public class ToggleButton extends Button {
    private final String onStr;
    private final String offStr;
    private final Runnable onEnable;
    private final Runnable onDisable;
    private final String prefix;

    private boolean optionEnabled;

    public void setOptionEnabled(boolean enabled) {
        optionEnabled = enabled;

        if (optionEnabled) {
            setText(onStr);
            onEnable.run();
        } else {
            setText(offStr);
            onDisable.run();
        }
    }

    public boolean getOptionEnabled() {
        return optionEnabled;
    }

    public ToggleButton(double x, double y, double sX, double sY, String optionName, Runnable onEnable, Runnable onDisable, boolean initialState) {
        this(x,y,sX,sY,optionName,onEnable,onDisable,initialState,null);
    }
    public ToggleButton(double x, double y, double sX, double sY, String optionName, Runnable onEnable, Runnable onDisable, boolean initialState, String hoverText) {
        super(x, y, sX, sY, "");
        prefix = optionName+": ";

        onStr = prefix+ScreenOptions.onText;
        offStr = prefix+ScreenOptions.offText;
        this.onEnable = onEnable;
        this.onDisable = onDisable;

        setOptionEnabled(initialState);

        if (hoverText != null) {
            setHoverText(hoverText);
            this.enableHover = true;
        }

        this.function = () -> setOptionEnabled(!optionEnabled);

        this.enabled = true;
    }
}

File: src/tools/important/tankslua/gui/ToggleButton.kt
package tools.important.tankslua.gui

import tanks.gui.Button
import tanks.gui.screen.ScreenOptions

class ToggleButton(
    x: Double,
    y: Double,
    width: Double,
    height: Double,
    val label: String,
    val onToggled: (Boolean) -> Unit,
    initiallyEnabled: Boolean = false,
) : Button(x, y, width, height, label) {
    private fun updateText() {
        val onOff = if (toggled) ScreenOptions.onText else ScreenOptions.offText
        setText("$label: $onOff")
    }

    var toggled: Boolean = initiallyEnabled
        set(newEnabled) {
            field = newEnabled
            updateText()
            onToggled(field)
        }

    init {
        updateText()
        function = Runnable { toggled = !toggled }
        enabled = true
    }
}